src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -497,9 +499,27 @@ public override string MetadataName
         }
 
         /// <summary>
-        /// If this type is a file-local type, returns the syntax tree where this type is visible. Otherwise, returns null.
+        /// If this type is a file-local type, returns an identifier for the file this type was declared in. Otherwise, returns null.
         /// </summary>
-        internal abstract SyntaxTree? AssociatedSyntaxTree { get; }
+        internal abstract FileIdentifier? AssociatedFileIdentifier { get; }
+
+        internal struct FileIdentifier
+        {
+            public ImmutableArray<byte> FilePathChecksum { get; init; }
+            public string DisplayFilePath { get; init; }
+        }
+
+        public static FileIdentifier CreateFileIdentifier(SyntaxTree tree)
+        {
+            using var sha256 = SHA256.Create();
+            var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(tree.FilePath));
```

Comment:
>GetBytes [](http://example.com/codeflow?start=56&length=8)

What happens when UTF8 encoding fails?

Reply:
Probably the encoding will throw an exception. I think which encoding we use here is not important, as long as we always use the same one to create FileIdentifiers. So maybe it would be better to pick an encoding which is compatible with any possible FilePath string.

Reply:
I've pushed a change to ensure that any UTF8 encoding failure results in an error being reported. I stuck with UTF8 as the encoding because portable PDBs use it for [document names](https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md#DocumentNameBlob) and it seems useful to enforce that a file-local type name can round-trip through a portable PDB.